Create form views - MP-36, MP-37, MP-38, MP-39 (feature/mp-36-39)

dominikmlynarczyk opened the pull request:
**Below is a description of the changes:**

**1. Changes to the model**
Based on our discussions, I have applied changes to the model, especially the ingredient model. In the documentation we considered storing a picture (`imagePath`), but it doesn't make much sense to keep a picture for an ingredient (the end user won't be taking a picture of let's say "sugar", come on...). 
In the case of ingredients, if you think that it is worth to leave it, I suggest leaving it as an optional field, and providing placeholder image.

Similarly for ingredient categories - this is included in the documentation, but it doesn't make much sense in my opinion. Maybe when the application will expand and we will have a warehouse this need will appear, but I would let it go for now. If anyone thinks that it should stay, please provide reasoning - remember that you have to define a separate endpoint for ingredient categories, because these categories are not displayed in the terminal.

As we discussed, I added an optional `tax` type for the Category that will define a fixed tax value for the category and this value should be able to be inherited by all products/dishes in the category.

I applied some other changes to model - especially look for a change in imagePath to image, because `<input>` with the `file` type returns a `FileList` object and this will be needed when retrieving data from the form.

**2. Problems**
I had a problem with typing `name` and `register` fields, when creating an abstraction for form fields. I used the type `any` and this should be fixed - please give me a suggestion on this issue.

I don't really know how to solve the issue of calculating the price for a dish. The documentation here contains a lot of inconsistent information, because there we still consider the weight of net and gross products. What do we want to do with this, given that the warehouse is not ready?

**3. Known issues**
The Select element gets a placeholder value when the parent component is re-rendered - the correct value is sent, but this is strictly a UI issue. 

Calculating pricing for article & dish is not working - consider it for MP-36 to MP-39 tickets to be only a mock view.

dominikmlynarczyk committed:
feat: Add template for form, add atom comp & small refactor

- Separate UploadIcon to atom component
- Change styles for Maciej's category view, to fit properly inside
  grid (to be entirely refactored later)
- Change "Add category button" to be rendered as link

dominikmlynarczyk committed:
feat(model): Update docs model & add utility input-types

- Add tax for categories (to be inherited by children)
- Remove unnecessary image for ingredient
- Change type for imagePath -> image (to be discussed)

dominikmlynarczyk committed:
chore(eslint): Ignore require-default-props rule

As we use TypeScript this rule doesn't make much sense

dominikmlynarczyk committed:
feat(menu): Add form molecule components (WIP)

- Components are not well-typed due to the problem of preparing the
  right type for the register method and the name property
- Select component gets placeholder value, whenever parent gets
  re-rendered
- FormGroupProductPrice is WIP
- FormGroupProductPrice is not reusable and strictly will only be
  usable in the menu section. This is not a bug/problem, but it is
  worth paying attention
- Separate component for dish pricing is not implemented yet

dominikmlynarczyk committed:
feat(menu): Add form menu views (with mock data)

mfurmanit commented on src/components/atoms/FileUploadIcon/index.tsx:
```diff
@@ -0,0 +1,20 @@
+const FileUploadIcon = () => (
+  <Icon
+    mx="auto"
+    boxSize={12}
```
That's the thing which I wanted to discuss. 
Should we use braces (`{}`) or quote marks (`""`) when defining single values for styles properties?
(in `boxSize` you use braces, in the others quotes)
I suggest using quotes for single values and of course braces for more robust styling (responsiveness).

dominikmlynarczyk replied:
Thank you for pointing this out. I changed the value passing from `{}` notation to single quote notation. It is worth noting, that there are exceptions where Chakra requires props to be passed using `{}` (e.g. in `FormGroupFileUpload` the `borderWidth` prop).

mfurmanit commented on src/model/documents/category.ts:
```diff
@@ -1,7 +1,8 @@
-  imagePath: string;
   name: string;
   parent?: Category;
+  tax?: string[];
```
That should be probably typed as `Tax`, not array of string (??).

dominikmlynarczyk replied:
I applied some changes and caught brain lag probably. Good for us that you noticed that.
Fixed.

mfurmanit commented on src/others/input-types.ts:
```diff
@@ -0,0 +1,23 @@
+export type InputTypes =
```
That should be `InputType` without ending `s`.
We don't write `CategorieS` or `TaxeS`.
It's not array, it's a type.

dominikmlynarczyk replied:
Indeed, sorry. Already fixed 🩹.

mfurmanit commented on src/routing.tsx:
```diff
@@ -1,6 +1,6 @@
 import Employees from 'views/Access/Employees';
 import Dashboard from 'views/Dashboard';
-import Categories from 'views/Menu/Categories';
+import Categories from 'components/Tables/Categories'; // To be changed later for proper view - views/Menu/Categories
```
Write it as `// TODO` please - this way IDE will be able to investigate such things easily in the future.

dominikmlynarczyk replied:
Added "TODO: " for future code clean up.

mfurmanit commented on src/views/Menu/DishForm.tsx:
```diff
@@ -1,5 +1,111 @@
+          name="image"
+          register={register}
+          errors={errors}
+          validation={{ required: 'To pole jest wymagane' }}
```
You can add default validation messages to components starting with `FormGroup` - here you use same message for every required field.

dominikmlynarczyk replied:
Extracted error message to const.

mfurmanit commented on src/components/Table/Toolbar.tsx:
```diff
@@ -37,7 +40,7 @@ export interface ToolbarProps {
   <Styles>
     {props.showButton && (
-      <Button size="md" onClick={props.onButtonClick}>
+      <Button as={RouterLink} to={ROUTE_MENU_FORMS.CATEGORY.path} size="md" onClick={props.onButtonClick}>
```
It should be passed in props, because it is generic component that will be used in many tables.

dominikmlynarczyk replied:
Extracted error message to const.

dominikmlynarczyk replied:
We resolved this issue and agreed that this change would be in Maciej's PR.

mfurmanit replied:
Yup, it is already there. 👍

mfurmanit commented on src/components/molecules/FormGroupFileUpload/index.tsx:
```diff
@@ -0,0 +1,66 @@
+  id: string;
+
+  // eslint-disable-next-line @typescript-eslint/no-explicit-any
+  name: FieldPath<any>;
```
As we discussed on Teams.
You need to make `props` and component generic by passing `<T>` which will evaluate to types like `Category`, `Product` and so on.

dominikmlynarczyk replied:
We have temporarily solved this issue by replacing interfaces with types. This has its limitations, one of the biggest current bugs is having to remove the parent property from the Category type due to the "circularly referenced type in TypeScript" error

mfurmanit reviewed (changes requested):
Well done! 👍 
I have couple things to resolve / fix / discuss.

dominikmlynarczyk committed:
refactor(menu): Use quotes for passing props to Chakra components

Some props such as borderWidth require numeric props passed as numbers,
so I left them untouched.

dominikmlynarczyk committed:
fix(model): Update Category interface & rename InputType

dominikmlynarczyk committed:
chore: Add "TODO" in routing for easier future updates / code clean up

dominikmlynarczyk committed:
refactor(menu): Extract required error message to const

dominikmlynarczyk committed:
feat(model): Use types instead of interfaces in the documents model

The change is due to the need for correct type handling
in React Hook Form

dominikmlynarczyk committed:
feat(model): Add MenuForm type to use it at the FormGroup component level

In the future, FormGroup components will use "wider" typing (more generic)

dominikmlynarczyk committed:
feat(menu): Update form views and FormGroup components with new typing

Temporary implementation until we find a better typing

dominikmlynarczyk committed:
feat: Comment categories component to make the current code work

Due to type issues Maciej needs to update Categories component

IMPORTANT: Update Categories component TableCategory typing
